File: src/oca/java/src/org/opennebula/client/vmgroup/VMGroup.java
/*******************************************************************************
 * Copyright 2002-2021, OpenNebula Project, OpenNebula Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package org.opennebula.client.vmgroup;


import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.PoolElement;
import org.w3c.dom.Node;

/**
 * This class represents an OpenNebula vmgroup.
 * It also offers static XML-RPC call wrappers.
 */
public class VMGroup extends PoolElement{

    private static final String METHOD_PREFIX   = "vmgroup.";
    private static final String ALLOCATE        = METHOD_PREFIX + "allocate";
    private static final String INFO            = METHOD_PREFIX + "info";
    private static final String UPDATE          = METHOD_PREFIX + "update";
    private static final String DELETE          = METHOD_PREFIX + "delete";
    private static final String CHOWN           = METHOD_PREFIX + "chown";
    private static final String CHMOD           = METHOD_PREFIX + "chmod";
    private static final String RENAME          = METHOD_PREFIX + "rename";
    private static final String LOCK            = METHOD_PREFIX + "lock";
    private static final String UNLOCK          = METHOD_PREFIX + "unlock";

    /**
     * Creates a new vmgroup representation.
     *
     * @param id The vmgroup id.
     * @param client XML-RPC Client.
     */
    public VMGroup(int id, Client client)
    {
        super(id, client);
    }

    /**
     * @see PoolElement
     */
    protected VMGroup(Node xmlElement, Client client)
    {
        super(xmlElement, client);
    }

    // =================================
    // Static XML-RPC methods
    // =================================

    /**
     * Allocates a new vmgroup in OpenNebula.
     *
     * @param client XML-RPC Client.
     * @param description A string containing the template
     * of the vmgroup.
     *
     * @return If successful the message contains the associated
     * id generated for this vmgroup.
     */
    public static OneResponse allocate(
            Client  client,
            String  description)
    {
        return client.call(ALLOCATE, description);
    }

    /**
     * Retrieves the information of the given vmgroup
     *
     * @param client XML-RPC Client.
     * @param id the id for the vmgroup to
     * retrieve the information from.
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse info(Client client, int id)
    {
        return client.call(INFO, id);
    }

    /**
     * Retrieves the information of the given VMGroup.
     *
     * @param client XML-RPC Client.
     * @param id The VMGroup id for the VMGroup to retrieve the information from
     * @param decrypt If true decrypt sensitive attributes
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse info(Client client, int id, boolean decrypt)
    {
        return client.call(INFO, id, decrypt);
    }

    /**
     * Deletes a vmgroup from OpenNebula.
     *
     * @param client XML-RPC Client.
     * @param id The id of the target vmgroup.
     * @return A encapsulated response.
     */
    public static OneResponse delete(Client client, int id)
    {
        return client.call(DELETE, id);
    }

    /**
     * Changes the owner/group
     *
     * @param client XML-RPC Client.
     * @param id The id of the target vmgroup.
     * @param uid The new owner user ID. Set it to -1 to leave the current one.
     * @param gid The new group ID. Set it to -1 to leave the current one.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse chown(Client client, int id, int uid, int gid)
    {
        return client.call(CHOWN, id, uid, gid);
    }

    /**
     * Changes the vmgroup permissions
     *
     * @param client XML-RPC Client.
     * @param id The id of the target vmgroup.
     * @param owner_u 1 to allow, 0 deny, -1 do not change
     * @param owner_m 1 to allow, 0 deny, -1 do not change
     * @param owner_a 1 to allow, 0 deny, -1 do not change
     * @param group_u 1 to allow, 0 deny, -1 do not change
     * @param group_m 1 to allow, 0 deny, -1 do not change
     * @param group_a 1 to allow, 0 deny, -1 do not change
     * @param other_u 1 to allow, 0 deny, -1 do not change
     * @param other_m 1 to allow, 0 deny, -1 do not change
     * @param other_a 1 to allow, 0 deny, -1 do not change
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse chmod(Client client, int id,
                                    int owner_u, int owner_m, int owner_a,
                                    int group_u, int group_m, int group_a,
                                    int other_u, int other_m, int other_a)
    {
        return chmod(client, CHMOD, id,
                owner_u, owner_m, owner_a,
                group_u, group_m, group_a,
                other_u, other_m, other_a);
    }

    /**
     * Changes the permissions
     *
     * @param client XML-RPC Client.
     * @param id The id of the target object.
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse chmod(Client client, int id, String octet)
    {
        return chmod(client, CHMOD, id, octet);
    }

    /**
     * Changes the permissions
     *
     * @param client XML-RPC Client.
     * @param id The id of the target object.
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse chmod(Client client, int id, int octet)
    {
        return chmod(client, CHMOD, id, octet);
    }

    /**
     * Replaces the vmgroup template contents.
     *
     * @param client XML-RPC Client.
     * @param id The id of the target vmgroup we want to modify.
     * @param new_template New template contents.
     * @param append True to append new attributes instead of replace the whole template
     * @return If successful the message contains the vmgroup id.
     */
    public static OneResponse update(Client client, int id, String new_template,
        boolean append)
    {
        return client.call(UPDATE, id, new_template, append ? 1 : 0);
    }

    /**
     * Renames this vmgroup
     *
     * @param client XML-RPC Client.
     * @param id The id of the target vmgroup.
     * @param name New name for the vmgroup.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse rename(Client client, int id, String name)
    {
        return client.call(RENAME, id, name);
    }

    /**
     * lock this vmgroup
     *
     * @param client XML-RPC Client.
     * @param id The id of the target vmgroup.
     * @param level Lock level.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse lock(Client client, int id, int level)
    {
        return client.call(LOCK, id, level);
    }

    /**
     * Unlock this vmgroup
     *
     * @param client XML-RPC Client.
     * @param id The id of the target vmgroup.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse unlock(Client client, int id)
    {
        return client.call(UNLOCK, id);
    }

    // =================================
    // Instanced object XML-RPC methods
    // =================================

    /**
     * Loads the xml representation of the vmgroup.
     * The info is also stored internally.
     *
     * @see VMGroup#info(Client, int)
     */
    public OneResponse info()
    {
        OneResponse response = info(client, id);
        super.processInfo(response);
        return response;
    }

    /**
     * Deletes the vmgroup from OpenNebula.
     *
     * @return A encapsulated response.
     */
    public OneResponse delete()
    {
        return delete(client, id);
    }

    /**
     * Changes the owner/group
     *
     * @param uid The new owner user ID. Set it to -1 to leave the current one.
     * @param gid The new group ID. Set it to -1 to leave the current one.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chown(int uid, int gid)
    {
        return chown(client, id, uid, gid);
    }

    /**
     * Changes the owner
     *
     * @param uid The new owner user ID.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chown(int uid)
    {
        return chown(uid, -1);
    }

    /**
     * Changes the group
     *
     * @param gid The new group ID.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chgrp(int gid)
    {
        return chown(-1, gid);
    }

    /**
     * Changes the vmgroup permissions
     *
     * @param owner_u 1 to allow, 0 deny, -1 do not change
     * @param owner_m 1 to allow, 0 deny, -1 do not change
     * @param owner_a 1 to allow, 0 deny, -1 do not change
     * @param group_u 1 to allow, 0 deny, -1 do not change
     * @param group_m 1 to allow, 0 deny, -1 do not change
     * @param group_a 1 to allow, 0 deny, -1 do not change
     * @param other_u 1 to allow, 0 deny, -1 do not change
     * @param other_m 1 to allow, 0 deny, -1 do not change
     * @param other_a 1 to allow, 0 deny, -1 do not change
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chmod(int owner_u, int owner_m, int owner_a,
                             int group_u, int group_m, int group_a,
                             int other_u, int other_m, int other_a)
    {
        return chmod(client, id,
                    owner_u, owner_m, owner_a,
                    group_u, group_m, group_a,
                    other_u, other_m, other_a);
    }

    /**
     * Changes the permissions
     *
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chmod(String octet)
    {
        return chmod(client, id, octet);
    }

    /**
     * Changes the permissions
     *
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chmod(int octet)
    {
        return chmod(client, id, octet);
    }

    /**
     * Replaces the vmgroup template contents.
     *
     * @param new_template New template contents.
     * @return If successful the message contains the vmgroup id.
     */
    public OneResponse update(String new_template)
    {
        return update(new_template, false);
    }

    /**
     * Replaces the vmgroup template contents.
     *
     * @param new_template New template contents.
     * @param append True to append new attributes instead of replace the whole template
     * @return If successful the message contains the vmgroup id.
     */
    public OneResponse update(String new_template, boolean append)
    {
        return update(client, id, new_template, append);
    }

    /**
     * Renames this vmgroup
     *
     * @param name New name for the vmgroup.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse rename(String name)
    {
        return rename(client, id, name);
    }

    /**
     * Lock this vmgroup
     *
     * @param level Lock level.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse lock(int level)
    {
        return lock(client, id, level);
    }

    /**
     * Unlock this vmgroup
     *
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse unlock()
    {
        return unlock(client, id);
    }

    // =================================
    // Helpers
    // =================================

}

File: src/oca/ruby/opennebula/vm_group.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

require 'opennebula/lockable_ext'
require 'opennebula/pool_element'

module OpenNebula
    class VMGroup < PoolElement
        #######################################################################
        # Constants and Class Methods
        #######################################################################
        VMGROUP_METHODS = {
            :allocate    => "vmgroup.allocate",
            :info        => "vmgroup.info",
            :update      => "vmgroup.update",
            :delete      => "vmgroup.delete",
            :chown       => "vmgroup.chown",
            :chmod       => "vmgroup.chmod",
            :rename      => "vmgroup.rename",
            :lock        => "vmgroup.lock",
            :unlock      => "vmgroup.unlock",
            :roleadd     => "vmgroup.roleadd",
            :roledelete  => "vmgroup.roledelete",
            :roleupdate  => "vmgroup.roleupdate"
        }

        # Creates a VMGroup description with just its identifier
        # this method should be used to create plain VMGroup objects.
        # @param pe_id [Integer] the id of the object
        def VMGroup.build_xml(pe_id=nil)
            if pe_id
                obj_xml = "<VM_GROUP><ID>#{pe_id}</ID></VM_GROUP>"
            else
                obj_xml = "<VM_GROUP></VM_GROUP>"
            end

            XMLElement.build_xml(obj_xml,'VM_GROUP')
        end

        # Class constructor
        def initialize(xml, client)
            LockableExt.make_lockable(self, VMGROUP_METHODS)

            super(xml,client)

            @client = client
        end

        #######################################################################
        # XML-RPC Methods for the VMGroup Object
        #######################################################################

        # Retrieves the information of the VMGroup.
        def info()
            super(VMGROUP_METHODS[:info], 'VM_GROUP')
        end

        alias_method :info!, :info

        # Allocates a new VMGroup in OpenNebula
        #
        # @param description [String] The contents of the VMGroup.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def allocate(description)
            super(VMGROUP_METHODS[:allocate], description)
        end

        # Deletes the SecurityGroup
        def delete()
            super(VMGROUP_METHODS[:delete])
        end

        # Replaces the vm group contents
        #
        # @param new_vmgroup [String] New vmgroup contents
        # @param append [true, false] True to append new attributes instead of
        #   replace the whole securitygroup
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def update(new_vmgroup, append=false)
            super(VMGROUP_METHODS[:update], new_vmgroup, append ? 1 : 0)
        end

        # Changes the owner/group
        #
        # @param uid [Integer] the new owner id. Set to -1 to leave the current one
        # @param gid [Integer] the new group id. Set to -1 to leave the current one
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chown(uid, gid)
            super(VMGROUP_METHODS[:chown], uid, gid)
        end

        # Changes the SecurityGroup permissions.
        #
        # @param octet [String] Permissions octed , e.g. 640
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chmod_octet(octet)
            super(VMGROUP_METHODS[:chmod], octet)
        end

        # Changes the SecurityGroup permissions.
        # Each [Integer] argument must be 1 to allow, 0 deny, -1 do not change
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chmod(owner_u, owner_m, owner_a, group_u, group_m, group_a, other_u,
                other_m, other_a)
            super(VMGROUP_METHODS[:chmod], owner_u, owner_m, owner_a, group_u,
                group_m, group_a, other_u, other_m, other_a)
        end

        # Renames this VMGroup
        #
        # @param name [String] New name for the VMGroup.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def rename(name)
            return call(VMGROUP_METHODS[:rename], @pe_id, name)
        end

        # Add role to VM Group
        #
        # @param template [String] String template for the new role
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def role_add(template)
            return call(VMGROUP_METHODS[:roleadd], @pe_id, template)
        end

        # Delete role from VM Group
        #
        # @param roleid [Integer] ID of the role to remove
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def role_delete(roleid)
            return call(VMGROUP_METHODS[:roledelete], @pe_id, roleid)
        end

        # Update VM Group role
        #
        # @param roleid [Integer] ID of the role to remove
        # @param template [String] String template with updated values
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def role_update(roleid, template)
            return call(VMGROUP_METHODS[:roleupdate], @pe_id, roleid, template)
        end

        #######################################################################
        # Helpers to get VMGroup information
        #######################################################################

        # Returns the group identifier
        # [return] _Integer_ the element's group ID
        def gid
            self['GID'].to_i
        end

        def owner_id
            self['UID'].to_i
        end

        # [return] _Array_ with the name of roles
        def role_names
            self.retrieve_elements('ROLES/ROLE/NAME')
        end
    end
end
